napari/components/viewer_model.py

```diff
@@ -714,15 +714,16 @@ def open_sample(
             raise KeyError(msg)
 
         if callable(data):
+            added = []
             for datum in data(**kwargs):
-                self._add_layer_from_data(*datum)
-        elif isinstance(data, (str, Path)):
-            self.open(data, plugin=reader_plugin)
-        else:
-            raise TypeError(
-                'Got unexpected type for sample '
-                f'({plugin!r}, {sample!r}): {type(data)}'
-            )
+                added.extend(self._add_layer_from_data(*datum))
+            return added
+        if isinstance(data, (str, Path)):
+            return self.open(data, plugin=reader_plugin)
+        raise TypeError(
+            'Got unexpected type for sample '
+            f'({plugin!r}, {sample!r}): {type(data)}'
+        )
```

Comment:
Generally I prefer if/elif/else, even when the returns make them unnecessary, because they are easier to read as blocks in English. ie I can skip between the clauses and understand the flow of the function. Would you mind reverting this? It's not relevant to the PR, is it?